draft-ietf-tls-esni.md

```diff
@@ -430,6 +430,37 @@ the ClientHelloOuter. The server then computes the digest of the reconstructed
 ClientHelloInner. If the digest does not equal OuterExtensions.inner_digest,
 then the server MUST abort the connection with a "decrypt_error" alert.
 
+# The "handshake_padding" Extension {#handshake-padding}
+
+Encrypting the ClientHelloInner may not be sufficient to protect sensitive
+parameters. In particular, the length of the ClientHelloInner, or the length of
+the backend server's response, may leak information about these parameters. The
+client and backend server can use the "handshake_padding" extension in order to
+mitigate this leakage.
+
+A new extension type is defined, which the client MAY include in its
+ClientHello. It MUST NOT offer this extension unless it also advertises support
+for TLS 1.3 or above.
+
+~~~
+    enum {
+       handshake_padding(0xff05), (65535)
+    } ExtensionType;
+~~~
+
+When solicited by the client, the server MUST ignore the "handshake_padding"
+extension if it negotiates TLS 1.2 or below. Otherwise, it MAY reply by
+including the extension in its EncryptedExtensions message.
```

Comment:
I was going to suggest we go ahead and stick it into the other messages too for the client Finished, but maybe that's better deferred until later. Though we'd have to explicitly say it's not allowed in any other "request" messages but we may define what it means later, whereas any "response" messages we forbid it from now will be stuck. So that's just a mess.

In particular, if we believe in compressed certificates staying as-is, any client certificates story will need client EncryptedExtensions. That said, at least in the client EncryptedExtensions formulation we're pondering for ALPS, the signal would be in the wrong place so it's a mess all around. (https://github.com/vasilvv/tls-alps/pull/5. It's possible we came up with the wrong formulation---I still need to email the about it.)

@ekr, do you have thoughts?

(FWIW, a Padding message would avoid all this because it wouldn't need to interact with how each message fits into extension request/response patterns. ;-) We'd need to communicate a single boolean in either direction and then everyone is free to pad to their heart's content anywhere.)

Reply:
Given that the client-side piece is orthogonal to ECH, I think it's probably best to defer that until we figure out the client EE/ALPS story.